### PR TITLE
[NP-6455] - DaoWAO Implementation

### PR DESCRIPTION
**Pull request by nanopay-arthur**
Ref: https://nanopay.atlassian.net/browse/NP-6455

credit to @KernelDeimos for DaoWAO implementation

### CODE REVIEW

**Comment by KernelDeimos** [comment]
A test for this would be helpful right now - there are gonna be a few tasks before we start using it for real

**Review by KernelDeimos — APPROVED** [review]

File: src/foam/u2/wizard/tests.jrl [hunk 1]
```diff
@@ -0,0 +1,103 @@
+p({
+  class:"foam.nanos.test.Test",
+  id:"DaoWAOTest",
+  description: "Unit test for DAO Wizardlet Access Object",
+  language: 0,
+  code:"""
+    async function execute(x) {
```
Comment (KernelDeimos) [inline src/foam/u2/wizard/tests.jrl:7]: I believe there's a way to do this in a js file but still register it in a journal

Comment (KernelDeimos) [reply src/foam/u2/wizard/tests.jrl:7]: as per discussion, no convenient way to do this right now (something that needs to be resolved)

File: src/foam/u2/wizard/DaoWAO.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
+/**
+ * @license
+ * Copyright 2022 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.u2.wizard',
+  name: 'DaoWAO',
```
Comment (kgrgreer) [inline src/foam/u2/wizard/DaoWAO.js]: By the coding-standard, it should be called DAOWAO. All letters of an acronym are treated like one letter, so when capitalized, all should be.

File: src/foam/u2/wizard/DaoWAO.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+          });
+          expr = expr ? this.DOT(expr, partialProperty) : partialProperty;
+        }
+        return expr;
+      }
+    },
+    {
+      class: 'Boolean',
+      name: 'disableFind',
+      value: false
```
Comment (kgrgreer) [inline src/foam/u2/wizard/DaoWAO.js]: Is the default default, so is redundant.

File: src/foam/u2/wizard/tests.jrl [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        
+        dw.data.email = 'francis@filth.com'
+        dw.data.userName = 'francis'
+        dw.data.group = 'sme'
+
+        resp = await dw.save();
+
+        if ( resp && resp.id ) {
+          test(true, 'successful DAOWAO save (no path), user id: ' + resp.id);
+        }
```
Comment (kgrgreer) [inline src/foam/u2/wizard/tests.jrl]: } else {

File: src/foam/u2/wizard/WAO.js [hunk 5]
```diff
@@ -8,6 +8,9 @@ foam.INTERFACE({
   package: 'foam.u2.wizard',
   name: 'WAO',
   label: 'Wizardlet Access Object',
+  todo: [
```
Comment (kgrgreer) [inline src/foam/u2/wizard/WAO.js:11]: I was unaware of this axiom ;-).

File: src/foam/u2/wizard/tests.jrl [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        else {
+          test(false, 'failed DAOWAO load (no path)');
+        }
+
+
+        // CASE: 3 - SAVE DATA WITH NESTED FOBJECT (WITH PATH)
+        dw = foam.u2.wizard.DAOWizardlet.create({ of: foam.nanos.crunch.UserCapabilityJunction, path: 'payload' }, x);
+        
+        ucj = foam.nanos.crunch.UserCapabilityJunction.create({ 
+                sourceId: 1348,
```
Comment (kgrgreer) [inline src/foam/u2/wizard/tests.jrl]: indentation looks off

File: src/foam/u2/wizard/DaoWAO.js (hunk 2 above)
Comment (KernelDeimos) [reply src/foam/u2/wizard/DaoWAO.js]: How would `DAOWAO` get correctly converted into camelCase or snake_case?

File: src/foam/nanos/crunch/example/capabilities.jrl [hunk 7]
```diff
@@ -2,6 +2,7 @@ p({
   "class": "foam.nanos.crunch.Capability",
   "id": "crunch.example",
   "name": "Example Capability",
+  "of": "foam.nanos.crunch.example.ExampleData",
   "enabled": true,
   "visible": false
 })
```
Comment (tharmaman) [inline src/foam/nanos/crunch/example/capabilities.jrl]:
```suggestion
})

```

File: src/foam/u2/wizard/DAOWizardlet.js [hunk 8]
```diff
@@ -23,26 +27,17 @@ foam.CLASS({
       }
     },
     {
-      name: 'daoKey',
       class: 'String',
-      expression: function (of) {
-        return foam.String.daoize(of.name);
-      }
+      documentation: `
+        Optional path used to specify the location of the desired wizardlet\'s data object that will
```
Comment (tharmaman) [inline src/foam/u2/wizard/DAOWizardlet.js]:
```suggestion
        Optional path used to specify the location of the desired wizardlet's data object that will
```

No need to escape the apostrophe when using backticks

File: src/foam/u2/wizard/DaoWAO.js (hunk 2 above)
Comment (tharmaman) [reply src/foam/u2/wizard/DaoWAO.js]: Think it would it end up being `daowao`

File: src/foam/u2/wizard/DaoWAO.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+      class: 'foam.dao.DAOProperty',
+      expression: function (daoKey) {
+        return this.__context__[daoKey];
+      }
+    },
+    {
+      class: 'String',
+      name: 'path',
+      documentation: `
+        Optional path used to specify the location of the desired wizardlet\'s data object that will
```
Comment (tharmaman) [inline src/foam/u2/wizard/DaoWAO.js]:
```suggestion
        Optional path used to specify the location of the desired wizardlet's data object that will
```

File: src/foam/u2/wizard/tests.jrl [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+          test(false, 'failed DAOWAO disable load');
+        }
+      } catch (e) {
+        console.error(e.message || e);
+      }
+    }
+
+    await execute(x);
+  """
+})
```
Comment (tharmaman) [inline src/foam/u2/wizard/tests.jrl]:
```suggestion
})

```

File: src/foam/u2/wizard/DaoWAO.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        console.error(e);
+        wizardlet.reportNetworkFailure(e);
+      })
+      .finally(() => {
+        wizardlet.loading = false;
+      });
+    },
+    async function load(wizardlet) {
+      if ( wizardlet.loading ) return;
+      if ( ! this.disableFind ) {
```
Comment (tharmaman) [inline src/foam/u2/wizard/DaoWAO.js:104]: Is the disableFind case for when we want to use the DAOWAO in a create case?

File: src/foam/u2/wizard/DaoWAO.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        lastObject = wizardlet;
+        lastKey = 'data';
+      } else if ( this.Property.isInstance(path) ) {
+        lastObject = wizardlet.data;
+        lastKey = path;
+      } else if ( this.Dot.isInstance(path) ) {
+        lastObject = path.arg1.f(obj);
+        lastKey = path.arg2.name;
+      } else {
+        throw new Error('unexpected case in DaoWAO.setValue');
```
Comment (tharmaman) [inline src/foam/u2/wizard/DaoWAO.js]:
```suggestion
        throw new Error('Unexpected case in DaoWAO.setValue');
```

File: src/foam/u2/wizard/DaoWAO.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    {
+      class: 'foam.mlang.ExprProperty',
+      name: 'path_',
+      expression: function(path) {
+        if ( typeof path !== 'string' || path == '' ) return path;
+        const parts = path.split('.');
+        let expr = null;
+
+        for ( let part of parts ) {
+          const partialProperty = foam.core.Property.create({
```
Comment (tharmaman) [inline src/foam/u2/wizard/DaoWAO.js:68]: using const and lets only eh?

**Review by tharmaman — APPROVED** [review]

File: src/foam/u2/wizard/WAO.js (hunk 5 above)
Comment (KernelDeimos) [reply src/foam/u2/wizard/WAO.js:11]: It's new; it's so new that it technically doesn't exist yet.

Comment (KernelDeimos) [reply src/foam/u2/wizard/WAO.js:11]: While we could add it to `foam.core.Model` or a documentation-related refinement, my thinking is to leave it in this purgatory intentionally and decide where to put that when it becomes useful.

File: src/foam/u2/wizard/DaoWAO.js (hunk 11 above)
Comment (nanopay-arthur) [reply src/foam/u2/wizard/DaoWAO.js:104]: yeah, when we want to restrict the DAOWAO for only the create case.

File: src/foam/u2/wizard/WAO.js (hunk 5 above)
Comment (kgrgreer) [reply src/foam/u2/wizard/WAO.js:11]: FOAM1 had issues: []

https://github.com/foam-framework/foam/blob/9e8500f149625aebbc5dd5eabd722b1225ce3ce1/core/mm1Model.js#L829

Which was basically a cross between an issue tracker and more structured TODOs. Then you just use git as your issue tracker and create a view of the ModelDAO for viewing issues. Then the issue gets updated in the same PR as the code that fixes it.

**Pull request merged**